File: Taller/TallerEnPython/Shaker.py
from archivo_utils import ArchivoUtils  # Asumiendo que guardaste el código anterior como archivo_utils.py

class Shaker:
    """
    Esta clase implementa el método de la sacudida o Shaker Sort.
    Autor original Java: leon
    """

    @staticmethod
    def shaker(arreglo):
        """
        El método que ordena mediante la sacudida.
        :param arreglo: La lista a ordenar.
        :return: La lista ya ordenada.
        """
        intercambios = 0
        comparaciones = 0
        n = len(arreglo)
        
        izq = 1
        der = n - 1
        k = n - 1

        while izq <= der:
            # Fase 1: Derecha a izquierda (mueve el menor al principio)
            for i in range(der, izq - 1, -1):
                comparaciones += 1
                if arreglo[i - 1] > arreglo[i]:
                    intercambios += 1
                    arreglo[i - 1], arreglo[i] = arreglo[i], arreglo[i - 1] # Swap elegante de Python
                    k = i
            
            izq = k + 1

            # Fase 2: Izquierda a derecha (mueve el mayor al final)
            for i in range(izq, der + 1):
                comparaciones += 1
                if arreglo[i - 1] > arreglo[i]:
                    intercambios += 1
                    arreglo[i - 1], arreglo[i] = arreglo[i], arreglo[i - 1]
                    k = i
            
            der = k - 1

        print(f"Número de intercambios: {intercambios}")
        print(f"Número de comparaciones: {comparaciones}")
        return arreglo

def main():
    # En Python, para pasar una función como parámetro (como el lambda en Java),
    # simplemente pasamos el nombre del método.
    
    ArchivoUtils.medir_tiempo("archivo_10k.txt", "Shaker", Shaker.shaker)
    ArchivoUtils.medir_tiempo("archivo_100k.txt", "Shaker", Shaker.shaker)
    # Cuidado: 1M en Shaker Sort (O(n²)) puede tardar mucho en Python
    # ArchivoUtils.medir_tiempo("archivo_1M.txt", "Shaker", Shaker.shaker)

    ArchivoUtils.exportar_csv("Shaker")
    ArchivoUtils.exportar_json("Shaker")

if __name__ == "__main__":
    main()

File: Taller/TallerEnJava/src/main/java/com/analisis/algoritmos/Shaker.java
package com.analisis.algoritmos;
/**
 * Esta clase implementa el método de la sacudida o Shaker Sort
 * @author leon
 *
 */
public class Shaker {
	
	public static void main(String[] args) {

		ArchivoUtils.medirTiempo("archivo_10k.txt", "Shaker", arr -> shaker(arr));
		ArchivoUtils.medirTiempo("archivo_100k.txt", "Shaker", arr -> shaker(arr));
		ArchivoUtils.medirTiempo("archivo_1M.txt", "Shaker", arr -> shaker(arr));
	
		ArchivoUtils.exportarCSV("Shaker");
		ArchivoUtils.exportarJSON("Shaker");
	}
	/**
	 * El método que ordena mediante la sacudida
	 * @param arreglo El arreglo a ordenar.
	 * @return El arreglo ya ordenado.
	 * 
	 * Fuente: https://gist.github.com/leonlipe/488e3e6c67811ae2e34086e38ae96ee7
	 *  
	 */
    public static int[] shaker(int[] arreglo){
        int intercambios = 0, comparaciones = 0;
        int i, izq, der, k, aux;

        izq = 1;
        der = arreglo.length - 1;
        k = arreglo.length - 1;

        while (izq <= der) {

            for (i = der; i >= izq; i--) {
                comparaciones++;
                if (arreglo[i - 1] > arreglo[i]) {
                    intercambios++;
                    aux = arreglo[i - 1];
                    arreglo[i - 1] = arreglo[i];
                    arreglo[i] = aux;
                    k = i;
                }
            }

            izq = k + 1;

            for (i = izq; i <= der; i++) {
                comparaciones++;
                if (arreglo[i - 1] > arreglo[i]) {
                    intercambios++;
                    aux = arreglo[i - 1];
                    arreglo[i - 1] = arreglo[i];
                    arreglo[i] = aux;
                    k = i;
                }
            }

            der = k - 1;
        }

        System.out.println("Numero de intercambios: " + intercambios);
        System.out.println("Numero de comparaciones: " + comparaciones);

        return arreglo;
    }
}